### ISSUE TITLE
variable bound by @ in pattern matches is not visible in where clauses

### ISSUE THREAD

**Issue by jhegedus42** [idris-lang/Idris-dev#2978]
The following does not type check:

```
data D = A | B | C

d2int : D->Int
d2int A =1  

printD:(d:D)->String
printD (d@A) = "a" ++ ( show $ f d)
  where 
   f:D->Int
   f A= d2int d
```

however the corresponding Haskell code works fine.

**Comment by jhegedus42** [comment]
Similarly : 
```
data D = A|B

g:D->Type
g A =  String
g B =  Nat


f: (d:D)-> g d 
f  A   =  "bla"
f (d@B)=  1
  where 
    h: g d
    h = 2
```

fails to type check too.

**Comment by ahmadsalim** [comment]
This is still an issue in Idris `0.12`

**Comment by BartAdv** [comment]
I did some poking and it looks like it's caused by the fact as patterns are just desugared: x@y = let x = y in .... This means the rhs contains the newly assigned names, but the where block does not.

**Comment by eraserhd** [comment]
@jhegedus42 `d` in your second example is in the argument position, so it becomes an implicit (also, `g` needs to be declared `total` to be evaluated by the type-checker).

You can't refer to `d` in a qualified way until surface-level syntax is implemented (see #3991) – and that likely won't cover `@` variables without more work – and it will be implicit if unqualified.  So, even if I *did* preserve the value, you couldn't access the value without new syntax.

(PR #4057 will have `d` bound on the right-hand side of functions, but not in type signatures.)

So this becomes a language design question: Should variables bound by pattern matching or `@` be available as values in type signatures in `where` clauses?  So far, we've decided no.  I think the *consistent* answer is yes.  I think either will surprise people sometimes, unfortunately.

@edwinb (Or other designer people), what do you think?
